datasources/src/main/java/org/opensearch/sql/datasources/utils/XContentParserUtils.java

```diff
@@ -133,4 +186,38 @@ public static XContentBuilder convertToXContent(DataSourceMetadata metadata) thr
     builder.endObject();
     return builder;
   }
+
+  /**
+   * Converts Map<String, Object> (partial DataSourceMetadata) to XContentBuilder.
+   *
+   * @param dataSourceData
+   * @return XContentBuilder {@link XContentBuilder}
+   * @throws Exception Exception.
+   */
+  public static XContentBuilder convertMapToXContent(Map<String, Object> dataSourceData)
+      throws Exception {
+
+    XContentBuilder builder = XContentFactory.jsonBuilder();
+    builder.startObject();
+    if (dataSourceData.containsKey(DESCRIPTION_FIELD)) {
+      builder.field(DESCRIPTION_FIELD, dataSourceData.get(DESCRIPTION_FIELD));
+    }
+    if (dataSourceData.containsKey(ALLOWED_ROLES_FIELD)) {
+      builder.field(
+          ALLOWED_ROLES_FIELD, ((List<String>) dataSourceData.get(ALLOWED_ROLES_FIELD)).toArray());
+    }
+    if (dataSourceData.containsKey(PROPERTIES_FIELD)) {
+      builder.startObject(PROPERTIES_FIELD);
+      for (Map.Entry<String, String> entry :
+          ((Map<String, String>) dataSourceData.get(PROPERTIES_FIELD)).entrySet()) {
+        builder.field(entry.getKey(), entry.getValue());
+      }
+      builder.endObject();
+    }
```

Comment:
This might need the whole Properties to be upserted - otherwise data might be lost, with the current implementation.